### PR TITLE
Add renderRates.js to visualize rates over time

### PR DESCRIPTION
**Pull request by zzmp**

### CODE REVIEW

**Comment by zzmp** [comment]
@jherico this one's for you, have fun.

**Comment by hifi-gustavo** [comment]
Build console output for commit: efc7db6dafba70e3966c04a568462d24b2d9e531 available at: https://gist.github.com/8ef2fb26cd4aaf54af7d8b0b99fd85b0<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.dmg

**Comment by zzmp** [comment]
See: https://app.asana.com/0/74051501495175/111185415544234.

#### Testing

Open up the new renderRates.js and make sure it stays in sync with existing stats, in `Developer > Stats`.

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -508,6 +508,26 @@ Q_GUI_EXPORT void qt_gl_set_global_share_context(QOpenGLContext *context);
 
 Setting::Handle<int> sessionRunTime{ "sessionRunTime", 0 };
 
+class Rates : public QObject {
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: coding standard says this should be in it's own file.

**Comment by hifi-gustavo** [comment]
Build console output for commit: b11d6ca1c6723d04703055ed08f25937e8dbf1ae available at: https://gist.github.com/8ef2fb26cd4aaf54af7d8b0b99fd85b0<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.dmg

**Comment by ZappoMan** [comment]
```
/data/jenkins/Jenkins/workspace/beta_pr-master/label/ubuntu/libraries/shared/src/shared/RateCounter.h: In constructor ‘constexpr RateCounter<>::RateCounter()’:
/data/jenkins/Jenkins/workspace/beta_pr-master/label/ubuntu/libraries/shared/src/shared/RateCounter.h:22:7: error: use of deleted function ‘std::atomic<_Tp>::atomic(const std::atomic<_Tp>&) [with _Tp = float]’
 class RateCounter {
       ^
In file included from /data/jenkins/Jenkins/workspace/beta_pr-master/label/ubuntu/libraries/plugins/src/plugins/DisplayPlugin.h:11:0,
                 from /data/jenkins/Jenkins/workspace/beta_pr-master/label/ubuntu/libraries/display-plugins/src/display-plugins/DisplayPlugin.h:9,
                 from /data/jenkins/Jenkins/workspace/beta_pr-master/label/ubuntu/libraries/display-plugins/src/display-plugins/OpenGLDisplayPlugin.h:10,
                 from /data/jenkins/Jenkins/workspace/beta_pr-master/label/ubuntu/libraries/display-plugins/src/display-plugins/OpenGLDisplayPlugin.cpp:8:
/usr/include/c++/4.8/atomic:169:7: error: declared here
       atomic(const atomic&) = delete;
       ^
/data/jenkins/Jenkins/workspace/beta_pr-master/label/ubuntu/libraries/display-plugins/src/display-plugins/OpenGLDisplayPlugin.cpp: In constructor ‘OpenGLDisplayPlugin::OpenGLDisplayPlugin()’:
/data/jenkins/Jenkins/workspace/beta_pr-master/label/ubuntu/libraries/display-plugins/src/display-plugins/OpenGLDisplayPlugin.cpp:206:42: note: synthesized method ‘constexpr RateCounter<>::RateCounter()’ first required here 
 OpenGLDisplayPlugin::OpenGLDisplayPlugin() {
                                          ^

```

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5919c79164abd23ce127054743b22cc104791b8a available at: https://gist.github.com/8ef2fb26cd4aaf54af7d8b0b99fd85b0<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.dmg

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -1390,6 +1391,7 @@ void Application::initializeUi() {
     rootContext->setContextProperty("Preferences", DependencyManager::get<Preferences>().data());
     rootContext->setContextProperty("AddressManager", DependencyManager::get<AddressManager>().data());
     rootContext->setContextProperty("FrameTimings", &_frameTimingsScriptingInterface);
+    rootContext->setContextProperty("Rates", new RatesScriptingInterface(this));
```
Comment (zzmp) [inline interface/src/Application.cpp]: A new object for each context. Alternatively, I could put this in the DependencyManager.

File: interface/src/Application.cpp [hunk 3]
```diff
@@ -4443,6 +4445,8 @@ void Application::registerScriptEngineWithApplicationServices(ScriptEngine* scri
     // AvatarManager has some custom types
     AvatarManager::registerMetaTypes(scriptEngine);
 
+    scriptEngine->registerGlobalObject("Rates", new RatesScriptingInterface(this));
```
Comment (zzmp) [inline interface/src/Application.cpp]: A new object for each engine. Alternatively, I could put this in the DependencyManager.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1fe12e70294a95535c1374cdd7c728aa3a66bdfe available at: https://gist.github.com/8ef2fb26cd4aaf54af7d8b0b99fd85b0<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 477d6f40e4bd1292ecdf4c4e6eaf0f636354c31f available at: https://gist.github.com/8ef2fb26cd4aaf54af7d8b0b99fd85b0<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ad225b3aed6fa465992b587e836a67da87ffea74 available at: https://gist.github.com/8ef2fb26cd4aaf54af7d8b0b99fd85b0<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2c7cf3ea8378271970971edd91db0404cc3cd7e3 available at: https://gist.github.com/8ef2fb26cd4aaf54af7d8b0b99fd85b0<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Alpha-PR7708.dmg

**Comment by kevinmthomas-carpool** [comment]
## QA Test
**PR Build:** 7708
**Testing Architecture:** 64-bit PC Desktop
**Testing OS:** Windows 10
**Testing Locations:** localhost
**Additional Test Results Requested:** Successful test.  New ```renderRates.js``` stays in sync with existing stats in ```Developer > Stats```.  Screen shot enclosed.
**Successful Logout/Crash:** Successful test.
capture

File: examples/utilities/render/rates.qml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+//
+import QtQuick 2.5
+import QtQuick.Controls 1.4
+import "../lib/plotperf"
+
+Item {
+    id: root 
+    anchors.fill: parent
+
+    property var caches: [ ["Present", "present"], ["Present", "present"], ["New", "newFrame"], ["Dropped", "dropped"], ["Simulation", "simulation"], ["Avatar", "avatar"] ]
```
Comment (zzmp) [inline examples/utilities/render/rates.qml]: One of these present rates should be render rate instead.

**Pull request merged**